content/blog/angular-extend-class/index.md

```diff
@@ -0,0 +1,757 @@
+
+```typescript
+@Component({
+  template: '',
+  selector: 'base-component'
+})
+class BaseComponent implements OnInit, OnDestroy {
+  height = window.innerHeight;
+  width = window.innerWidth;
+
+  // This needs to be an arrow function, otherwise `this` will bind to the Window
+  // For more, see: https://twitter.com/crutchcorn/status/1530104879271645184
+  onResize = () => {
+    this.height = window.innerHeight;
+    this.width = window.innerWidth;
+  }
+
+  ngOnInit() {
+    window.addEventListener('resize', this.onResize);
+  }
+
+  ngOnDestroy() {
+    window.removeEventListener('resize', this.onResize);
+  }
+}
+
+@Component({
+  selector: 'app-root',
+  template: `
+    <p>The window is {{height}}px high and {{width}}px wide</p>
+  `,
+})
+class AppComponent extends BaseComponent {
+}
+```
+
+This solves the error and now `AppComponent` tracks resizing as-expected!
+
+You'll notice, however, that while `BaseComponent` does have the `implements` keyword, the `AppComponent` does not. While it's seemingly not a _requirement_ to have the `implements` keyword on `AppComponent` in modern versions of Angular, it's still highly suggested.
+
+```typescript
+@Component({
+  selector: 'app-root',
+  template: `
+    <p>The window is {{height}}px high and {{width}}px wide</p>
+  `,
+})
+class AppComponent extends BaseComponent implements OnInit, OnDestroy {
+}
+```
+
+This is because it's easier to glance at the `AppComponent` code and see what lifecycle methods are used in the extended class or not.
+
+# Simplifying Base Component Usage by using `@Injectable` {#injectable}
+
+While our `BaseComponent` is extendible now, there's a new frustration that's arose as a result of using the `@Component`: We just registered a new component that can be accidentally used in another component's template.
+
+For example, if we had a template that looked like:
+
+```html
+<base-component></base-component>
+```
+
+We wouldn't get a compiler error, but would have a loose bit of code running needlessly. Ideally we'd like to have `BaseComponent` still able to use lifecycle methods without registering a new template tag.
+
+Fortunately, that's possible, as of Angular 9; simply remove `BaseComponent`'s `@Component` `selector` property and it won't register a new tag.
+
+```typescript
+@Component({
+  template: ''
+})
+class BaseComponent implements OnInit, OnDestroy {
+  // ...
+}
+```
+
+That solves one problem, but still leaves one present with using `@Component`: you must add a declaration of the `BaseComponent` into an `NgModule`. Otherwise, you'll end up with the following error during compilation:
+
+```
+BaseComponent is not declared in any Angular module 
+```
+
+Luckily, [since Angular 10 you can now use `@Injectable` to declare your `BaseComponent` instead](https://angular.io/guide/migration-injectable). This sidesteps the problem because `Injectable`s do not need to be declared:
```

Comment:
Now wait a minute! This is starting to feel very hacky. Services do not have an OnInit hook! If this works it's because this class is incidentally extended by a component sub-class.